river/stats/cov.py

```diff
@@ -39,17 +47,130 @@ def __init__(self, ddof=1):
         self.mean_y = mean.Mean()
         self.cov = 0
 
-    def update(self, x, y):
+    def update(self, x, y, w=1.):
         dx = x - self.mean_x.get()
-        self.mean_x.update(x)
-        self.mean_y.update(y)
+        self.mean_x.update(x, w)
+        self.mean_y.update(y, w)
         dy = y - self.mean_y.get()
-        self.cov += (dx * dy - self.cov) / max(1, self.mean_x.n - self.ddof)
+        self.cov += w * (dx * dy - self.cov) / max(1, self.mean_x.n - self.ddof)
         return self
 
     def get(self):
         return self.cov
 
+    def __add__(self, other):
+        result = Cov(ddof=self.ddof)
+
+        # Update mean estimates
+        result.mean_x = self.mean_x + other.mean_x
+        result.mean_y = self.mean_y + other.mean_y
+
+        if result.mean_x.n <= result.ddof:
+            return result
+
+        # Scale factors
+        scale_a = self.mean_x.n - self.ddof
+        scale_b = other.mean_x.n - other.ddof
+
+        # Scale the covariances
+        result.cov = scale_a * self.cov + scale_b * other.cov
+        # Apply correction factor
+        result.cov += (
+            (self.mean_x.get() - other.mean_x.get()) * (self.mean_y.get() - other.mean_y.get())
+            * ((self.mean_x.n * other.mean_x.n) / result.mean_x.n)
+        )
+        # Reapply scale
+        result.cov /= (result.mean_x.n - result.ddof)
```

Comment:
Can't you do something like the following?

```py
result = copy.deepcopy(self)
result += other
```

I didn't think about it much but it might save you some lines of code.

Reply:
Yeap! Nice catch. It simplifies the code and reduces the amount of repeated logic :D